pkg/controller/instance/plan_execution_engine.go

```diff
@@ -1,332 +1,228 @@
-			configs["phaseName"] = phase.Name
-			configs["stepName"] = step.Name
-			configs["StepNumber"] = strconv.FormatInt(int64(j), 10)
-			var resources []runtime.Object
-			stepState, _ := getStepFromStatus(step.Name, phaseState)
-
-			engine := kudoengine.New()
-			for _, t := range step.Tasks {
-				if taskSpec, ok := plan.tasks[t]; ok {
-					resourcesAsString := make(map[string]string)
-
-					for _, res := range taskSpec.Resources {
-						if resource, ok := plan.templates[res]; ok {
-							templatedYaml, err := engine.Render(resource, configs)
-							if err != nil {
-								newState.Status = v1alpha1.ExecutionFatalError
-								phaseState.Status = v1alpha1.ExecutionFatalError
-								stepState.Status = v1alpha1.ExecutionFatalError
-
-								err := errwrap.Wrap(err, "error expanding template")
-								log.Print(err)
-								return nil, &executionError{err, true, nil}
-							}
-							resourcesAsString[res] = templatedYaml
-						} else {
-							newState.Status = v1alpha1.ExecutionFatalError
-							phaseState.Status = v1alpha1.ExecutionFatalError
-							stepState.Status = v1alpha1.ExecutionFatalError
-
-							err := fmt.Errorf("PlanExecution: Error finding resource named %v for operator version %v", res, meta.operatorVersionName)
-							log.Print(err)
-							return nil, &executionError{err, true, nil}
-						}
-					}
-
-					resourcesWithConventions, err := renderer.applyConventionsToTemplates(resourcesAsString, ExecutionMetadata{
-						engineMetadata: *meta,
-						planName:       plan.name,
-						phaseName:      phase.Name,
-						stepName:       step.Name,
-					})
-
-					if err != nil {
-						newState.Status = v1alpha1.ErrorStatus
-						phaseState.Status = v1alpha1.ErrorStatus
-						stepState.Status = v1alpha1.ErrorStatus
-
-						log.Printf("Error creating Kubernetes objects from step %v in phase %v of plan %v and instance %s/%s: %v", step.Name, phase.Name, plan.name, meta.instanceNamespace, meta.instanceName, err)
-						return nil, &executionError{err, false, nil}
-					}
-					resources = append(resources, resourcesWithConventions...)
-				} else {
-					newState.Status = v1alpha1.ErrorStatus
-					phaseState.Status = v1alpha1.ErrorStatus
-					stepState.Status = v1alpha1.ErrorStatus
-
-					err := fmt.Errorf("Error finding task named %s for operator version %s", taskSpec, meta.operatorVersionName)
-					log.Print(err)
-					return nil, &executionError{err, false, nil}
-				}
-			}
-
-			perStepResources[step.Name] = resources
+func getOrCreateStepStatus(stepName string, phaseStatus *v1alpha1.PhaseStatus) *v1alpha1.StepStatus {
+	for i, p := range phaseStatus.Steps {
+		if p.Name == stepName {
+			return &phaseStatus.Steps[i]
 		}
 	}
 
-	return result, nil
-}
+	log.Printf("PlanExecution: creating missing (!) step %s status for the phase status: %+v", stepName, phaseStatus)
+	stepStatus := &v1alpha1.StepStatus{Name: stepName}
+	phaseStatus.Steps = append(phaseStatus.Steps, *stepStatus)
```

Comment:
won't appending like this make the steps in the status report not ordered right? I think that matter for serial execution... I am not sure we should ever append like this, we should always just create the status based on the OV blueprint which is what we do in instance_types now. I would kind of lean on just failing here for now when the status does not exist and then in the next PR move toward creating the status as a whole here (basically moving the logic from instance types)

Reply:
i like the rationale here, :+1: if there's not some overarching reason.

Reply:
My rationale for it was: it wasn't enough reason to fail the execution if we can just re-create that part of the `Status`. But you're right about the wrong ordering...  Ok, added error handling here.